api/service/service.go

```diff
@@ -1,126 +1,111 @@
-}
+	functionName := req.Func
+	params := req.Params
+	auth := req.Auth
 
-func (s *Service) ServiceRequest(req Payload) {
-	// function := req.Function
-	// params := req.Params
-	// auth := req.Auth
-	// if (!auth || Object.keys(auth) == 0) auth = null;
+	if auth == nil || len(auth) == 0 {
+		auth = nil
+	}
+
+	s.mux.RLock()
+	funcInfo := s.function
+	s.mux.RUnlock()
 
+	function, ok := funcInfo[functionName]
+	if !ok {
+		s.client.Send(websocket.ServiceRequest, model.FunctionsPayload{ID: req.ID, Error: "No function registered on the service"})
+		return
+	}
+	response, err := function(params, auth)
+	if err != nil {
+		log.Println(err)
+	}
+	s.client.Send(websocket.ServiceRequest, model.FunctionsPayload{ID: req.ID, Params: response})
 }
 
-// Start is used to start the particular service (is Blocking)
-func (s *Service) Start() {
-
-	con := s.config.Transport.GetWebsockConn()
-	for {
-		if con != nil {
-			log.Println("Connected to Space Cloud")
-
-			c := make(chan *Payload, 10)
-			go func() {
-				for payload := range c {
-					if err := con.WriteJSON(payload); err != nil {
-						log.Println("Error in sendin data  ", err)
-					}
-				}
-			}()
-			c <- &Payload{Service: s.service, Type: utils.TypeServiceRegister, Id: s.id, Project: s.config.Project, Token: s.config.Token}
-			for {
-				in := Payload{}
-				err := con.ReadJSON(&in)
-				if err != nil {
-					log.Println("Error reading json.", err)
-				}
-
-				if in.Type == utils.TypeServiceRegister {
-					if in.Id == s.id {
-						temp := make(map[string]bool)
-						json.Unmarshal(in.Params, &temp)
-						if temp["ack"] {
-							log.Println("Service registered")
-						} else {
-							close(c)
-							panic("Could Not Register Service")
-						}
-					}
-				} else if in.Type == utils.TypeServiceRequest {
-					function, found := s.funcs[in.Function]
-					if found {
-						go function(&model.Message{in.Params}, &model.Message{in.Auth}, func(kind string, res interface{}) {
-							if kind != "response" {
-								close(c)
-								panic("Type must be 'response'")
-							} else {
-								answer, err := json.Marshal(res)
-								if err != nil {
-									log.Println("Could not marshal the result", res)
-									c <- &Payload{Service: s.service, Type: utils.TypeServiceRequest, Id: in.Id, Error: "Error parsing the result"}
-								} else {
-									c <- &Payload{Service: s.service, Type: utils.TypeServiceRequest, Id: in.Id, Params: answer}
-								}
-							}
-						})
-					} else {
-						c <- &Payload{Service: s.service, Type: utils.TypeServiceRequest, Id: in.Id, Error: "Function Not Registered"}
-					}
-				}
-			}
-			close(c)
-		} else {
-			log.Println("Not connected. Attempting to Reconnect...")
+func (s *Service) Start() error {
+	s.client.RegisterCallback(websocket.ServiceRequest, s.serviceRequest)
+
+	v, err := s.client.Request(websocket.ServiceRegister, model.ServiceRegisterRequest{Service: s.service, Project: s.options.Project, Token: s.options.Token,})
```

Comment:
make a function names serviceRegister in helpers to take care of this. Code isn't DRY right now.

Reply:
done